opencxl/cxl/device/cxl_type1_device.py

```diff
@@ -201,11 +219,92 @@ def _init_device(
     def get_reg_vals(self):
         return self._cxl_io_manager.get_cfg_reg_vals()
 
-    async def cxl_cache_readline(self, addr: int) -> int:
-        pass
+    async def get_next_cqid(self) -> int:
+        cqid: int
+        async with self._cqid_assign_lock:
+            cqid = next(self._cqid_gen)
+        return cqid
+
+    async def cxl_cache_readline(self, addr: int, cqid: Optional[int] = None) -> Future[int]:
+        loop = get_running_loop()
+        fut_data = loop.create_future()  # ugly solution
+
+        def _listen_check_set_fut(pckt: CxlCacheH2DRspPacket | CxlCacheH2DDataPacket):
+            # for now, ignore the 32B transfer case.
+            if isinstance(pckt, CxlCacheH2DDataPacket):
+                # received a data packet
+                # since we're ignoring the 32B transfer case,
+                # we can just assume this data packet contains everything we want
+                fut_data.set_result(pckt.data)
+                raise CancelledError
+            if (
+                pckt.h2drsp_header.cache_opcode != CXL_CACHE_H2DRSP_OPCODE.GO
+                or pckt.h2drsp_header.rsp_data
+                in (CXL_CACHE_H2DRSP_CACHE_STATE.INVALID, CXL_CACHE_H2DRSP_CACHE_STATE.ERROR)
+            ):
+                raise CancelledError  # terminate the listener and free the cqid
+
+        await self._cxl_cache_manager.send_d2h_req_rdown(addr)
```

Comment:
I agree that scoped functions are ugly. However, the definitions of these scoped functions rely on parameters passed into their containing functions. If I were to move these scoped functions out, I'd have to define factory functions that would also be methods of cxl_type1_device.py. Let me know if this is acceptable.

Reply:
I meant: Instead of
```
def foo():
  a = 3
  def foo2(num):
    num = 4
  foo2(a)
```

```
def foo():
  def foo2(num):
    num = 4
  a = 3
  foo2(a)
```

Reply:
OHH, makes sense. Will do!

Reply:
Oh shoot I just realized that might be a bit difficult to do. The callback has to receive a reference to the future in order to be able to set it. The future has to be referenced outside the callback so that the enclosing function can return it. I suppose `cxl_cache_readlines` can create the futures and pass them into `readline` by reference...